Added material-ui AutoComplete

noisecapella opened the pull request:
#### What are the relevant tickets?
Fixes #277 


#### What's this PR do?
Adds material-ui so we can use the AutoComplete component to replace our select fields. Also updates `boundStateSelectField` to call `boundSelectField` instead of implementing duplicate behavior

#### Where should the reviewer start?

#### How should this be manually tested?
Go to the profile and edit the gender field, or the state/country fields. You should notice these behaviors:
 - the autocomplete now only matches prefixes (case insensitive)
 - when user is not actively typing in the autocomplete, it should always show the currently selected item or a blank string.
 - when the user types some text and presses enter and they have something in the autocomplete, the first item in the autocomplete menu will become the selected item
 - when the user types some text which doesn't match anything and presses enter, it should revert back to whatever was previously selected
 - when user clicks on a selected item, it should become the new selected item
 - user should not be able to clear the field

Weird behaviors:
 - when you press down on the keyboard the autocomplete loses focus and the text field reverts to the currently selected item. Ideally we should let the user choose options by pressing the down key and pressing enter, but I don't know how to distinguish a loss of focus to go to the menu vs a loss of focus to go to another part of the form
#### Any background context you want to provide?

#### Screenshots (if appropriate)

#### What GIF best describes this PR or how it makes you feel?

noisecapella commented:
Needs a CSS update to make the text color consistent with react-mdl

aliceriot commented on static/js/dashboard.js:
```diff
@@ -24,6 +21,10 @@ if (!Object.entries) {
 
+// material-ui requirement
+let injectTapEventPlugin = require("react-tap-event-plugin");
+injectTapEventPlugin();
```
is this required just for the material-ui component to work? I'm just wondering if we're worrying about the usability of MM on a touch device at all.

noisecapella replied:
I'm not sure if it's required for AutoComplete but I wasn't able to click a SelectField without this code for some reason

aliceriot replied:
haha, sounds like a good idea to include it then :P

aliceriot commented:
Functionally this seems pretty good, and I think it's a huge step up in appearance from the `react-select`.

Not sure exactly what to do about the point you bring up (w/r/t typing in text which doesn't correspond to one of the options). There's another confusing behavior there, which is if you already have a value entered for a field. Then it's possible to type in some text which doesn't match any option, as which point the field looks exactly the same as if you had picked a valid choice, and you aren't prevented from proceeding to the next point. This is confusing because it appears to the user that whatever they've just typed in ('asdfasdfasdf' or something) has actually been saved for, say, 'Country', but actually 'Country' is still just set to whatever value it was before.

Anyway, it looks like we can add a bunch of callbacks (`onInput` and so on) so we may be able to hack the behavior we want in? not sure.

aliceriot commented:
Anyway, code looks good to me. If you're happy with the behavior for now I say :+1:

giocalitri commented:
it needs a rebase

noisecapella commented:
Ok, I've changed a lot of stuff so you may want to go through it all again. By the way I removed `boundMonthField` and `boundYearField` in favor of `boundDateField`. We will add back a `boundMonthYearField` in #276

aliceriot commented on static/js/components/PersonalTab.js:
```diff
@@ -45,30 +46,34 @@ class PersonalTab extends ProfileTab {
-      {this.boundTextField(["preferred_name"], "Preferred name")}<br />
-      {this.boundSelectField(['gender'], 'Gender', this.genderOptions)}<br />
-      {this.boundSelectField(
+    return <Grid>
```
you may want to add `className="profile-tab-grid"` here, so the spacing and width is in line with the other tabs.

aliceriot commented on static/js/dashboard_routes.js:
```diff
@@ -27,15 +27,15 @@ export function makeDashboardRoutes(browserHistory, store, onRouteUpdate) {
+              <Route path="privacy" component={PrivacyTab} />
             </Route>
-            <Route path="/terms_of_service" component={TermsOfServicePage}/>
+            <Route path="/terms_of_service" component={TermsOfServicePage} />
```
I agree that `<Foo prop="asdf" />` is more readable than `<Foo prop="asdf"/>`, is there an eslint rule we can turn on to enforce this?

aliceriot replied:
I took a look, it doesn't seem like there is one :(

aliceriot commented:
I'm noticing that the material-ui autocomplete and textfields are noticeably slower than what we're replacing. is this something you've noticed too @noisecapella ?

noisecapella commented:
Yes I noticed this too. I'm hoping it will be faster in the minimized version but I have no idea

noisecapella commented:
Do you have your console open while messing around with these fields?

For AutoComplete there is a `menuCloseDelay` field which makes things slower for some reason. We can also turn off animation where we can

aliceriot commented:
yeah I had the console open, I'll try with it closed. my computer is probably a little bit slower that yours.

aliceriot commented:
ah ok yeah closing the console makes it a lot faster, not a problem.

aliceriot commented:
yeah, we may want to turn off some of the animation. the animation when you select something in the autocomplete is a bit strange, it takes longer than it should to get out of the way.

aliceriot commented on static/js/util/profile_edit.js:
```diff
@@ -30,12 +32,16 @@ export function boundTextField(keySet, label) {
   };
+
+  // fullWidth means set width to 100% instead of 256px. This lets us use the
+  // Grid and Cell to manage its size
```
:100:

noisecapella replied:
One of many quirks of material-ui

aliceriot commented on static/js/util/profile_edit.js:
```diff
@@ -49,36 +55,66 @@ export function boundSelectField(keySet, label, options) {
-    _.set(clone, keySet, value? value.value : '');
+    let toStore;
+    if (option.value !== null && option.value !== undefined) {
+      toStore = option.value.props.value;
```
so here we're pulling the option value out of the `MenuItem` component we declare below (in `convertOption`). is there any reason we couldn't also stick a `value` prop on the return value of that function? it might make the code here a little simpler.

noisecapella replied:
I'd rather not since we can't call it `option.value` and I can't think of a better name for it. It also means I'd have to update the current unit tests so that `onChange` can make use of it

aliceriot replied:
ok, in that case do you think it would be wise to change this line to:

```js
toStore = _.get(option, ['value', 'props', 'value'])
```

for defensive-ness' sake?

aliceriot commented on static/js/util/profile_edit.js:
```diff
@@ -49,36 +55,66 @@ export function boundSelectField(keySet, label, options) {
+  let selectedValue = _.get(profile, keySet);
+  let selectedOption;
+  for (let option of options) {
+    if (option.value === selectedValue) {
```
you might be able to use `Array.find` here, something like

```js
options.find(option => option.value === _.get(profile, keySet))
```

aliceriot commented on static/js/util/profile_edit.js:
```diff
@@ -49,36 +55,66 @@ export function boundSelectField(keySet, label, options) {
+    } else {
+      return {};
+    }
+  };
```
this seems like we have to do some juggling to get the way we currently declare select options (`[{value: 3, label: "Foo" }]`) to work with this field. Would it make more sense to change how they're declared to match the requirements of this new component? (e.g. in `ProfileTab.js`

noisecapella replied:
I feel like the current way is cleaner since the `<MenuItem ...` stuff is hidden away in this function, and we don't need to duplicate `option.label` like we do here.

aliceriot commented on static/js/util/profile_edit.js:
```diff
@@ -89,14 +125,7 @@ export function boundStateSelectField(stateKeySet, countryKeySet, label) {
-      onChange={onChange} />
-    <span className="validation-error-text">{_.get(errors, stateKeySet)}</span>
-  </div>;
+  return mockedBoundSelectField.call(this, stateKeySet, label, options);
```
this is slick!

aliceriot commented:
couple comments!

noisecapella commented:
I updated the behavior to set the field to whatever the currently selected value is when the focus leaves the field. Still need to clean it up and write tests for it, it's getting pretty messy

aliceriot commented:
cool, I'll take another look after standup!

noisecapella commented:
Should we allow the user to clear the autocomplete in any way? I feel like it's simpler if we don't allow them to, and I can't think of how the UI would work, but I'm not sure if there's a use case we want to allow.

noisecapella commented:
So the current behavior should look like this:
 - the autocomplete now only matches prefixes (case insensitive)
 - when user is not actively typing in the autocomplete, it should always show the currently selected item or a blank string.
 - when the user types some text and presses enter and they have something in the autocomplete, the first item in the autocomplete menu will become the selected item
 - when the user types some text which doesn't match anything and presses enter, it should revert back to whatever was previously selected
 - when user clicks on a selected item, it should become the new selected item
 - user should not be able to clear the field

Weird behaviors:
 - when you press down on the keyboard the autocomplete loses focus and the text field reverts to the currently selected item. Ideally we should let the user choose options by pressing the down key and pressing enter, but I don't know how to distinguish a loss of focus to go to the menu vs a loss of focus to go to another part of the form

noisecapella commented:
Tests are now in place!

aliceriot commented:
Just going through your bullet points above:

- [x] the autocomplete now only matches prefixes (case insensitive)
- [ ] when user is not actively typing in the autocomplete, it should always show the currently selected item or a blank string.
- [ ] when the user types some text and presses enter and they have something in the autocomplete, the first item in the autocomplete menu will become the selected item
- [ ] when the user types some text which doesn't match anything and presses enter, it should revert back to whatever was previously selected
- [ ] when user clicks on a selected item, it should become the new selected item
user should not be able to clear the field

noisecapella commented:
We can change it to always show all options (do no filtering) but this breaks the behavior where Enter picks the first item in the list. I suppose we could have Enter do a comparison that the text in the text field matches exactly, then choose that item, else revert to what was already selected.

For the hint text do you have an idea where to put that in the UI?

aliceriot commented:
hmm not sure, it's kind of hard to make things like that unobtrusive while still being clear. maybe we could conditionally modify the `floatingTextLabel` prop?

noisecapella commented:
We don't have a way to tell whether the user opened the menu via mouse click or by entering text, at least not without digging deeper into AutoComplete, which is a little unpalatable

noisecapella commented:
By the way review comments should be addressed now

aliceriot commented:
cool, I just looked over the code and it LGTM. how do we feel about merging what we have, and then coming back around sometime in the future to deal with some of the small UX edge cases?

noisecapella commented:
Sounds good

aliceriot commented:
:+1:
